### ISSUE TITLE
SSH Environment does not match app environment

### ISSUE THREAD

**Issue by Jonty** [cloudfoundry/diego-ssh#29]
When an app runs it has it's `HOME` set to `/home/vcap/app` and all of the files in `/home/vcap/app/.profile.d` are sourced in order to set up the expected environment.

When a user SSH's to the same app container, the `HOME` is set to `/home/vcap` and none of the `.profile.d` files are sourced.

This means that tenants are unable to load installed libraries, run management commands, or really do anything with the app in the container unless they fix this first. Tenants can be instructed to run `export HOME=/home/vcap/app; for f in ~/.profile.d/*; do source $f; done` when they first connect, but I do not feel this should be necessary.

**Comment by caod123** [comment]
There is some context as to why this is the case here: http://cf-dev.70369.x6.nabble.com/cf-dev-In-Diego-maybe-DEA-also-why-is-HOME-home-vcap-app-and-not-home-vcap-td4802.html

**Comment by jvshahid** [comment]
tl;dr; the [launcher](https://github.com/cloudfoundry/buildpackapplifecycle/blob/master/launcher/main.go) is responsible for setting the home env variable and sourcing in files in `.profiles.d`. we thought about using the launcher to start the ssh process. the tricky part is ensuring that the `.profiles.d` files are idempotent and won't be destructive to execute multiple times. since we cannot guarantee that, we held back on making this change. we are open for suggestions though

**Comment by Jonty** [comment]
For reference, the HOME var is set using the [diego SSH daemon](https://github.com/cloudfoundry/diego-ssh/blob/3ce14ee4e6e32cee2c2d34fbacb3529603ca58d6/handlers/session_channel_handler.go#L391) by copying it from the existing environment the daemon is running in.

Are you aware of any buildpacks where the .profiles.d files are not idempotent, or is this a theoretical concern?

**Comment by Jonty** [comment]
Update: I had a cursory glance over all the core buildpacks and couldn't find any that had non-idempotent `.profile.d`'s, most of them don't use it at all. I presume the concern is that they might be supplied by tenants?

**Comment by jvshahid** [comment]
that's correct, it is just a theoretical concern since there's no way to enforce that restriction on custom buildpacks.

**Comment by Jonty** [comment]
@jvshahid Would an alternative solution be to simply document that `.profile.d` scripts must be idempotent? This assumes that no known existing buildpacks exhibit that behaviour.

**Comment by ematpl** [comment]
Thanks for asking about this, @Jonty. That solution sounds like a behavioral restriction for buildpacks that would be most appropriate for @sclevine, the PM of the  Buildpacks, to address. It would likely also be appropriate to apply the same restriction to the optional app-provided `.profile` file that will be sourced after any `.profile.d` files, as that too may set up environment important to the deployed CF application process. Given the general guideline that CF attempts to be compatible with Heroku buildpacks, that may not be a reasonable restriction to enforce given the [currently documented contract for `.profile.d` and `.profile` scripts](https://devcenter.heroku.com/articles/buildpack-api#profile-d-scripts).

In the meantime, this environmental discrepancy is noted in the CF documentation at http://docs.cloudfoundry.org/devguide/deploy-apps/ssh-apps.html#ssh-env, along with manual steps to take to set up the environment to match that of the app processes. Also, although it does depend on current implementation details of how applications are run in containers, if you are certain for a particular app that it is safe to run the buildpack and app profile scripts repeatedly, you can run `cf ssh APP_NAME -t -c '/tmp/lifecycle/launcher app bash ""'` to start an interactive session with the same starting environment as the app processes.

More broadly, it sounds like we would want the capability to run one or more commands with the environment variables captured from an earlier stage of setup. We could use that capability to run the app start command and the SSH server simultaneously with the environment set up by the profile scripts. I have some initial ideas about how we might be able to achieve that through a receipt file inside the container filesystem. I'll discuss the feasibility of that approach with the Diego team and with the Greenhouse team, as it would be important for the same invocation flow to work on Windows cells as well as on Linux cells.

Thanks,
Eric

**Comment by Jonty** [comment]
Thanks @ematpl, please do keep this issue updated with any progress.

I noticed the docs you linked to were slightly incorrect and would lead to partial environment setup for some buildpacks, I've filed a PR to fix them: https://github.com/cloudfoundry/docs-dev-guide/pull/152

**Comment by sunjayBhatia** [comment]
you can now get the app environment by running `/tmp/lifecycle/shell` from within your app ssh session
